### PR TITLE
Kotlin-wasm changes to support UiBroker

### PR DESCRIPTION
**Pull request by sjmiles**
changes copied from other PR (#3646), minus confusing whitespace changes

### CODE REVIEW

File: src/runtime/wasm.ts [hunk 1]
```diff
@@ -667,6 +680,23 @@ export class WasmParticle extends Particle {
...
+  // so just go straight to output
+  output(content) {
+    this.container.apiPort.Output(this, content);
+  }
+
+  // render request call-back from wasm
+  onRenderOutput(templatePtr: WasmAddress, modelPtr: WasmAddress) {
+    const content: Content = {templateName: 'default'};
+    content.template = this.container.read(templatePtr);
+    content.model = new StringDecoder().decodeDictionary(this.container.read(modelPtr));
```
Comment (alxrsngrtn) [inline src/runtime/wasm.ts]: `StringDecoder.decodeDictionary(...)` should be sufficient.

Comment (sjmiles) [reply src/runtime/wasm.ts]: I used the patterns in `renderImpl` and `serviceRequest` ... is it wrong in those places also?

Comment (mykmartin) [reply src/runtime/wasm.ts]: No, StringDecoder sets and consumes a member field while decoding.

File: src/runtime/wasm.ts [hunk 2]
```diff
@@ -526,6 +527,18 @@ export class WasmParticle extends Particle {
...
     this.container.register(this, this.innerParticle);
+    // TODO(sjmiles): probably too soon: we need to render at least once, but we may have handle
+    // work pending. @shans says: if the particle has readable handles, onHandleUpdate is guaranteed
+    // to be called, otherwise we need `renderOutput` manually. Need to optimize this across all
+    // particle bases.
+    setTimeout(() => this.renderOutput(), 100);
+  }
+
+  renderOutput() {
+    // TODO(sjmiles): not yet implemented in CPP
```
Comment (mykmartin) [inline src/runtime/wasm.ts]: To stub this in CPP you need a function in the "JS-to-WASM" API section in arcs.cc:

```
EMSCRIPTEN_KEEPALIVE
void renderOutput(Particle* particle) {
}
```

File: src/runtime/wasm.ts [hunk 3]
```diff
@@ -526,6 +527,18 @@ export class WasmParticle extends Particle {
     }
     this.innerParticle = this.exports[fn]();
     this.container.register(this, this.innerParticle);
+    // TODO(sjmiles): probably too soon: we need to render at least once, but we may have handle
+    // work pending. @shans says: if the particle has readable handles, onHandleUpdate is guaranteed
+    // to be called, otherwise we need `renderOutput` manually. Need to optimize this across all
+    // particle bases.
+    setTimeout(() => this.renderOutput(), 100);
```
Comment (mykmartin) [inline src/runtime/wasm.ts]: We definitely don't want this long term, but if we need it to move forward that's fine. Ultimately I think there should be a lifecycle method called when all readable handles are ready, or immediately if there are none.

Comment (sjmiles) [reply src/runtime/wasm.ts]: Yes, this was just a hack to work around the absence of that method. In the meantime, I got the clarification from Shane about `onHandleUpdate` guarantee. This needs repairs in various Particle impls (assuming we aren't getting the new lifecycle method right away), so I didn't want to block this PR on it.

File: src/runtime/wasm.ts (hunk 2 above)
Comment (sjmiles) [reply src/runtime/wasm.ts]: My tack was to not touch the CPP until I/we actually implement `renderOutput`.

File: src/runtime/wasm.ts (hunk 3 above)
Comment (mykmartin) [reply src/runtime/wasm.ts]: Ok, sounds good.

File: src/wasm/kotlin/src/arcs/Particle.kt [hunk 4]
```diff
@@ -52,15 +52,20 @@ abstract class Particle : WasmObject() {
...
-      if (sendModel) {
-        val dict = populateModel(slotName)
-        model = StringEncoder.encodeDictionary(dict)
-      }
+    fun renderOutput() {
+      log("renderOutput")
+      var slotName = ""
+      val template = getTemplate(slotName)
+      val dict = populateModel(slotName)
+      var model = StringEncoder.encodeDictionary(dict)
```
Comment (alxrsngrtn) [inline src/wasm/kotlin/src/arcs/Particle.kt]: I think these can all be `val`s.

File: src/runtime/wasm.ts (hunk 1 above)
Comment (mykmartin) [reply src/runtime/wasm.ts]: Alex: look at the code. `decodeDictionary()` sets `this.str`, which means it needs an instance, which means `new StringDecoder().decodeDictionary` is correct.

File: src/wasm/kotlin/src/arcs/Particle.kt (hunk 4 above)
Comment (sjmiles) [reply src/wasm/kotlin/src/arcs/Particle.kt]: Doh, I didn't even notice those keywords were different. Interesting design decision for Kotlin, lol.

**Pull request merged**